[Merged by Bors] - feat(Topology/Order): add `Dense.exists_seq_strict{Mono/Anti}_tendsto`

wwylele opened the pull request:
Add `Dense.exists_seq_strict{Mono/Anti}_tendsto{'}` that restricts the sequence to a dense subset. This is useful for e.g. finding a monotone rational sequence approaching a real number.

Co-authored-by: Jireh Loreaux <loreaujy@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

j-loreaux reviewed (commented):
I think it would be nice to provide a `ℚ`-specific version too, since this isn't just a subset of `ℝ`. This works:

```lean
theorem Real.exists_seq_rat_strictMono_tendsto (x : ℝ) :
    ∃ u : ℕ → ℚ, StrictMono u ∧ (∀ n, u n < x) ∧ Tendsto (u · : ℕ → ℝ) atTop (𝓝 x) := by
  obtain ⟨u, hu₁, hu₂, hu₃⟩ := Rat.denseRange_cast.exists_seq_strictMono_tendsto x
  refine ⟨fun n ↦ (hu₂ n).2.choose, ?_, ?_⟩
  · simpa [StrictMono, ← (Rat.cast_strictMono (K := ℝ)).lt_iff_lt, fun n ↦ (hu₂ n).2.choose_spec]
  · aesop (add simp (fun n ↦ (hu₂ n).2.choose_spec))
```

wwylele commented on Mathlib/Topology/Order/IsLUB.lean:
```diff
@@ -142,6 +142,27 @@ theorem IsGLB.mem_of_isClosed {a : α} {s : Set α} (ha : IsGLB s a) (hs : s.Non
+protected theorem Set.Subset.isGLB_congr {α : Type*} [TopologicalSpace α] [Preorder α]
+    [ClosedIciTopology α] {s t : Set α} (hst : s ⊆ t) (hts : t ⊆ closure s) {x : α} :
+    IsGLB s x ↔ IsGLB t x :=
+  Set.Subset.isLUB_congr (α := αᵒᵈ) hst hts
```
This particular one doesn't seem to work for me, as `Set.Subset` name space doesn't recognize `hst : s ⊆ t` as a valid leading parameter. I can reproduce this problem

```
import Mathlib.Data.Set.Defs

theorem Set.Subset.theoremA {s t : Set α} (hst : Set.Subset s t) : False := sorry

/-
replace theoremA with the following definition, theoremB breaks with:
invalid field notation, function 'theoremA' does not have argument with type (Set.Subset ...) that can be used, it must be explicit or implicit with a unique name
-/ 
--theorem Set.Subset.theoremA {s t : Set α} (hst : s ⊆ t) : False := sorry

theorem Set.Subset.theoremB {s t : Set α} (hst : s ⊆ t) : False := hst.theoremA
```

Perhaps `Set.Subset` and `⊆` are somehow different, so this doesn't work?

wwylele replied:
I did a search in mathlib and it seems that there are few theorems in Set.Subset, namely only the special ones like `trans`, `refl` etc., which seems to get special treatment. I can't get normal theorems in `Set.Subset` working with dot syntax, so I renamed these to long descriptive name instead.

j-loreaux replied:
Sorry, that's my fault. I was thinking of `HasSubset.Subset` which *would* allow dot notation, but that's not a good name. The ones you provided are fine.

Indeed, `⊆` is `HasSubset.Subset`, and the definition of this notation for `Set α` is `Set.Subset`, which is why you had the observed behavior.

wwylele commented:
Thanks! I have addressed most comments except one that I commented above

I have a local version of `Real.exists_seq_rat_strictMono_tendsto` but I was having a hard time making it mathlib-ready. Thanks for writing a short an nice proof!

I'd like to also learn more about `aesop` and mathlib's policy on it. I am still figuring out when exactly it will just work, and whether mathlib has any readability/performance concern on it. I will ask more on zulip, but if you happen to have a  pointer, it would be appreciated.

j-loreaux commented:
@urkud I'm happy with this, but I had far to big a hand in it to merge myself. Can you have a look?

wwylele commented:
@urkud have you got a chance to take a look?

urkud commented on Mathlib/Topology/Order/IsLUB.lean:
```diff
@@ -193,6 +214,23 @@ theorem exists_seq_tendsto_sSup {α : Type*} [ConditionallyCompleteLinearOrder 
   rcases (isLUB_csSup hS hS').exists_seq_monotone_tendsto hS with ⟨u, hu⟩
   exact ⟨u, hu.1, hu.2.2⟩
 
+theorem Dense.exists_seq_strictMono_tendsto' [DenselyOrdered α] [FirstCountableTopology α]
```
I know that we already use primed lemmas in this file, but let's not increase the technical debt and use a more descriptive modifier, e.g.,
```suggestion
theorem Dense.exists_seq_strictMono_tendsto_of_lt [DenselyOrdered α] [FirstCountableTopology α]
```

urkud commented on Mathlib/Topology/Order/IsLUB.lean:
```diff
@@ -193,6 +214,23 @@ theorem exists_seq_tendsto_sSup {α : Type*} [ConditionallyCompleteLinearOrder 
+    exact ⟨z, mem_inter hzx hyz⟩
+  have hx : IsLUB (Ioo y x ∩ s) x := hs.isLUB_inter_iff isOpen_Ioo |>.mpr <| isLUB_Ioo hy
+  apply hx.exists_seq_strictMono_tendsto_of_not_mem (by aesop) hnonempty |>.imp
+  aesop
```
This looks like a useful generalization of you `Real`/`Rat` lemma.
```suggestion
  aesop

theorem DenseRange.exists_seq_strictMono_tendsto_of_lt {β : Type*} [LinearOrder β]
    [DenselyOrdered α] [FirstCountableTopology α] {f : β → α} {x y : α} (hf : DenseRange f)
    (hmono : Monotone f) (hlt : y < x) :
    ∃ u : ℕ → β, StrictMono u ∧ (∀ n, f (u n) ∈ Ioo y x) ∧ Tendsto (f ∘ u) atTop (𝓝 x) := by
  rcases Dense.exists_seq_strictMono_tendsto_of_lt hf hlt with ⟨u, hu, huyx, huf, hlim⟩
  choose v hv using huf
  obtain rfl : f ∘ v = u := funext hv
  exact ⟨v, fun a b hlt ↦ hmono.reflect_lt <| hu hlt, huyx, hlim⟩
```

urkud commented:
I'm sorry for the delay. Please tag me after the next update, and I'll try to promptly review it.

wwylele commented:
@urkud Thanks for the review! I have addressed the comments. The proof of `exists_seq_strictMono_tendsto_of_lt` you provided for some reason didn't compile as-is for me, so I manually separated ` huyx` and `huf` there.

urkud commented:
Thanks! :tada:
bors merge
